### PR TITLE
feat(cq): re-working cq for new endpoints

### PR DESCRIPTION
**Pull request by thomasyopes**
Ref: #1040

### Dependencies

- upstream: none
- downstream: https://github.com/metriport/metriport/pull/3100

### Description

- consolidated getting CQ orgs, and remove the "currently active"
- allowing facility / org updates on the DB only with no HIE update incase DB gets out of sync
- aligning the hie PUT logic to update DB *then* HIE similar to the medical PUT routes 
- add new fhir directory API + parsing to support new directory

Outstanding:
- [ ] replicate CQ logic to CW
- [ ] Release CQ SDK

### Testing

- Local
  - [ ] test all CQ directory functions
- Staging
  - [ ] test all CQ directory functions
- Sandbox
  - [ ] N/A
- Production
  - [ ] test all CQ directory functions

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/api/src/external/commonwell/command/create-or-update-cw-organization.ts
Comment (RamilGaripov) [inline packages/api/src/external/commonwell/command/create-or-update-cw-organization.ts:1]: why the change here? if anything goes wrong during the call to CW, don't we want to NOT update the status to active?

File: packages/api/src/external/carequality/shared.ts [hunk 1]
```diff
@@ -171,20 +174,53 @@ export const cqOrgActiveSchema = z.object({
...
+
+export async function getCqOrg(
+  cq: CarequalityManagementAPI,
+  oid: string
+): Promise<CQOrganization | undefined> {
+  const { log } = out(`CQ getCqOrg - CQ Org OID ${oid}`);
+
+  try {
+    const orgExistsActive = await cq.listOrganizations({ oid, active: true });
+    const orgExistsInActive = await cq.listOrganizations({ oid, active: false });
```
Comment (RamilGaripov) [inline packages/api/src/external/carequality/shared.ts:202]: nit: `Inactive` instead of `InActive`

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts [hunk 2]
```diff
@@ -1,98 +1,79 @@
...
+import { CQOrgDetails, getCqAddress, getParsedCqOrgOrFail, getCqOrg } from "../../shared";
 import { metriportIntermediaryOid, metriportOid } from "./create-or-update-cq-facility";
 import { metriportEmail as metriportEmailForCq } from "../../constants";
-import { metriportCompanyDetails } from "@metriport/shared";
 
 const cq = makeCarequalityManagementAPI();
 
 export async function createOrUpdateCQOrganization(
-  orgDetails: CQOrgDetails,
-  active: boolean
```
Comment (RamilGaripov) [inline packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts:20]: why are we removing the param for `active`? IIRC, we use this to deactivate cx org on CQ

File: packages/api/src/external/carequality/shared.ts (hunk 1 above)
Comment (thomasyopes) [reply packages/api/src/external/carequality/shared.ts]: No longer relevant.

File: packages/api/src/external/carequality/organization-conversion.ts
Comment (leite08) [inline packages/api/src/external/carequality/organization-conversion.ts:1]: OOC: why moving them into `./gateway/index.ts`?

File: packages/api/src/external/carequality/organization-fhir-template.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+function getEndpoint(urnOid: string, urlType: ChannelUrl, url: string): Endpoint {
+  const channelType = urlType === XCPD_STRING ? "ihe-xcpd" : "ihe-xca";
+  const channelName =
+    urlType === XCPD_STRING
+      ? "Patient Discovery"
+      : urlType === XCA_DQ_STRING
+      ? "Query for Documents"
+      : "Retrieve Documents";
```
Comment (leite08) [inline packages/api/src/external/carequality/organization-fhir-template.ts:195]: Just to confirm, it can be only one of the three options?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 4]
```diff
@@ -92,16 +93,18 @@ router.post(
   upload.single("file"),
   asyncHandler(async (req: Request, res: Response) => {
     const file = req.file;
-    if (!file) {
-      throw new BadRequestError("File must be provided");
-    }
-    const bundle = JSON.parse(file.buffer.toString());
+    if (!file) throw new BadRequestError("File must be provided");
+    const bundle = JSON.parse(file.buffer.toString()) as Bundle;
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:97]: I know we were not doing this before, but don't we want to validate the bundle?

We have a shell for this on `parseFhirBundle()`, and could look into improving it by calling `validateFhirEntries()`.

Defs not a blocker, but could at least map the improvement if it becomes a scope creep.

File: packages/api/src/routes/medical/internal-cq.ts [hunk 5]
```diff
@@ -92,16 +93,18 @@ router.post(
...
 
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const orgs = bundle.map((e: any) => e.resource.Organization);
+    const orgs: Organization[] = (bundle.entry as Organization[]) ?? [];
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:100]: the cast at the right already indicates the type, so no need to specify it on the left of the assignment

File: packages/api/src/routes/medical/internal-cq.ts [hunk 6]
```diff
@@ -127,25 +130,13 @@ router.get(
   requestLogger,
   asyncHandler(async (req: Request, res: Response) => {
     if (Config.isSandbox()) return res.sendStatus(httpStatus.NOT_IMPLEMENTED);
-    const cq = makeCarequalityManagementAPI();
+    const cq = makeCarequalityManagementAPIFhir();
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:133]: ditto: try to keep the original Factory design pattern

File: packages/api/src/routes/medical/internal-cq.ts [hunk 7]
```diff
@@ -162,22 +153,22 @@ router.get(
...
-      const facility = await getFaciltiyByOidOrFail({ cxId, id: facilityId, oid });
-      cqOrg = await getParsedCqOrgOrFail(cq, oid, facility.cqActive);
+      await getFaciltiyByOidOrFail({ cxId, id: facilityId, oid });
+      org = await getCqOrgOrFail(cq, oid);
     } else {
-      const org = await getOrganizationByOidOrFail({ cxId, oid });
-      cqOrg = await getParsedCqOrgOrFail(cq, oid, org.cqActive);
+      await getOrganizationByOidOrFail({ cxId, oid });
+      org = await getCqOrgOrFail(cq, oid);
     }
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:169]: nit: could make `org` a const with something like

```ts
    const [, org] = await Promise.all([
      facilityId
        ? getFaciltiyByOidOrFail({ cxId, id: facilityId, oid })
        : getOrganizationByOidOrFail({ cxId, oid }),
      getCqOrgOrFail(cq, oid),
    ]);
```

File: packages/api/src/routes/medical/internal-facility.ts [hunk 8]
```diff
@@ -33,6 +33,7 @@ router.put(
   requestLogger,
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getUUIDFrom("query", req, "cxId").orFail();
+    const skipHie = getFromQueryAsBoolean("skipHie", req);
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-facility.ts:36]: Where is this being called/used?

File: packages/api/src/routes/medical/internal-facility.ts [hunk 9]
```diff
@@ -71,18 +70,17 @@ router.put(
     const syncInHie = await verifyCxItVendorAccess(cxId, false);
     // TODO Move to external/hie https://github.com/metriport/metriport-internal/issues/1940
     // CAREQUALITY
-    if (syncInHie && facility.cqApproved) {
+    if (syncInHie && facility.cqApproved && !skipHie) {
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-facility.ts:73]: We should prob combine both `syncInHie` and `skipHie` in a single one, prob `syncInHie` being initialized considering the value of `skipHie`?

File: packages/api/src/routes/medical/internal-organization.ts [hunk 10]
```diff
@@ -56,41 +57,36 @@ router.put(
...
+      await getOrganizationOrFail({ cxId, id: orgDetails.id });
       org = await updateOrganization({ id: orgDetails.id, ...organizationCreate });
     } else {
       org = await createOrganization(organizationCreate);
     }
     const syncInHie = await verifyCxProviderAccess(cxId, false);
     // TODO Move to external/hie https://github.com/metriport/metriport-internal/issues/1940
     // CAREQUALITY
-    if (syncInHie && org.cqApproved) {
+    if (syncInHie && org.cqApproved && !skipHie) {
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-organization.ts:69]: ditto: combine the flags

File: packages/carequality-sdk/src/client/carequality-fhir.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+  /**
+   * Registers an organization with the Carequality directory.
+   *
+   * @param org string containing the organization resource (in XML format)
+   * @returns an XML string containing an OperationOutcome resource - see Carequality documentation for details - https://carequality.org/healthcare-directory/OperationOutcome-create-success-example2.xml.html
+   */
+  async registerOrganization(org: Organization): Promise<Organization> {
+    const query = new URLSearchParams();
+    query.append("apikey", this.apiKey);
```
Comment (leite08) [inline packages/carequality-sdk/src/client/carequality-fhir.ts:179]: This is the only way to send the API Key? This is not the best practice, as URLs and parameters get logged. If we can, we should prob send that on the header.

File: packages/carequality-sdk/src/client/carequality-fhir.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    if (count < 1 || count > MAX_COUNT) {
+      throw new Error(`Count value must be between 1 and ${MAX_COUNT}`);
+    }
+
+    const query = new URLSearchParams();
+    query.append("apikey", this.apiKey);
+    query.append("_count", count.toString());
+    start !== undefined && query.append("_start", start.toString());
+    oid !== undefined && query.append("_id", oid);
+    active !== undefined && query.append("active", active.toString());
```
Comment (leite08) [inline packages/carequality-sdk/src/client/carequality-fhir.ts:163]: nit: use `!=` so its more flexible with `null` and clients using pure JS

File: packages/carequality-sdk/src/client/carequality-fhir-api.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+export interface CarequalityManagementAPIFhir {
+  listOrganizations({
+    count,
+    start,
+    oid,
+    active,
+  }: ListOrganizationsParams): Promise<Organization[]>;
+  registerOrganization(org: Organization): Promise<Organization>;
+  updateOrganization({ org, oid }: UpdateOrganizationParams): Promise<Organization>;
+}
```
[the remark this answers is not in the document]
Comment (thomasyopes) [reply packages/carequality-sdk/src/client/carequality-fhir-api.ts:24]: The return types are different so unclear we can reuse the interface... but I'll see

File: packages/carequality-sdk/src/client/carequality-fhir.ts (hunk 12 above)
Comment (thomasyopes) [reply packages/carequality-sdk/src/client/carequality-fhir.ts]: Done.

File: packages/api/src/routes/medical/internal-cq.ts (hunk 4 above)
Comment (thomasyopes) [reply packages/api/src/routes/medical/internal-cq.ts:96]: Yea... is there maybe something from medplum so I don't need to do it myself?

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts [hunk 14]
```diff
@@ -75,13 +50,13 @@ export async function updateCQOrganization(
 }
 
 export async function registerOrganization(
-  cq: CarequalityManagementAPI,
+  cq: CarequalityManagementAPIFhir,
```
[the remark this answers is not in the document]
Comment (thomasyopes) [reply packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts]: The return types are different so not clear to me how helpful a higher up interface would be...

File: packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts [hunk 15]
```diff
@@ -1,14 +1,19 @@
+import { MetriportError } from "@metriport/shared";
+import { Organization, Endpoint } from "@medplum/fhirtypes";
 import { ORG_POSITION, TRANSACTION_URL } from "@metriport/carequality-sdk/common/util";
 import { Address } from "@metriport/carequality-sdk/models/address";
 import { Contained } from "@metriport/carequality-sdk/models/contained";
-import { ManagingOrganization, Organization } from "@metriport/carequality-sdk/models/organization";
+import {
+  ManagingOrganization,
+  Organization as OrganziationLegacy,
```
[the remark this answers is not in the document]
Comment (thomasyopes) [reply packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts]: No longer relevant.

File: packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts [hunk 16]
```diff
@@ -67,6 +72,51 @@ export function parseCQDirectoryEntries(orgsInput: Organization[]): CQDirectoryE
...
+    lon: lon,
+    point,
+    addressLine,
+    city,
+    state,
+    zip: postalCode,
+    managingOrganization: parentOrgOid,
+    managingOrganizationId: parentOrgOid,
+    active,
+    lastUpdatedAtCQ: "TODO",
```
[the remark this answers is not in the document]
Comment (thomasyopes) [reply packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts]: No longer relevant.

File: packages/api/src/external/carequality/command/cq-directory/rebuild-cq-directory.ts [hunk 17]
```diff
@@ -26,21 +26,29 @@ const sequelize = initDbPool(dbCreds, {
...
 
+  let currentPosition = 0;
+  let isDone = false;
   try {
     await createTempCQDirectoryTable();
     while (!isDone) {
       try {
         const orgs = await cq.listOrganizations({ start: currentPosition, count: BATCH_SIZE });
-        if (orgs.length < BATCH_SIZE) isDone = true; // if CQ directory returns less than BATCH_SIZE number of orgs, that means we've hit the end
+        if (orgs.length < BATCH_SIZE) isDone = true;
```
[the remark this answers is not in the document]
Comment (thomasyopes) [reply packages/api/src/external/carequality/command/cq-directory/rebuild-cq-directory.ts:45]: The comment is just telling us how the while loop works, which is clear in my mind... it's not related to `listOrganizations()`

File: packages/api/src/external/carequality/organization-conversion.ts
Comment (thomasyopes) [reply packages/api/src/external/carequality/organization-conversion.ts:1]: Only place their used

File: packages/api/src/external/carequality/organization-fhir-template.ts (hunk 3 above)
Comment (thomasyopes) [reply packages/api/src/external/carequality/organization-fhir-template.ts]: This is the old logic, but yes I guess

File: packages/api/src/routes/medical/internal-cq.ts (hunk 7 above)
Comment (thomasyopes) [reply packages/api/src/routes/medical/internal-cq.ts]: No longer relevant.

File: packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts [hunk 18]
```diff
@@ -18,29 +22,31 @@ const XCA_DR_STRING = "ITI-39";
...
+
+  const active = org.active;
+  if (active == undefined) {
+    throw new MetriportError("Missing active on CQ Org", undefined, { org: stringify(org) });
+  }
 
-  const lastUpdatedAtCQ = org.meta?.lastUpdated;
-  if (!lastUpdatedAtCQ) return undefined;
+  if (!org.meta?.lastUpdated) log("Missing lastUpdated at CQ Org, using current timestamp");
+  const lastUpdatedAtCQ = org.meta?.lastUpdated ?? buildDayjs().toISOString();
```
Comment (thomasyopes) [inline packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts:37]: not 100% sure about this move... we can discuss

File: packages/api/src/external/commonwell/command/cw-organization/create-or-update-cw-organization.ts [hunk 19]
```diff
@@ -65,19 +66,21 @@ export async function createCWOrganization(
...
 ): Promise<CwOrgDetails> {
-  const { log, debug } = out(`CW ccreateCWOrganization - CW Org OID ${orgDetails.oid}`);
+  const { log, debug } = out(`CW createCWOrganization - CW Org OID ${orgDetails.oid}`);
   const commonwellOrg = await getOrganzationCwTemplate(orgDetails);
   const commonWell = makeCommonWellAPI(Config.getCWMemberOrgName(), Config.getCWMemberOID());
 
   try {
     const respCreate = await commonWell.createOrg(metriportQueryMeta, commonwellOrg);
-    debug(`resp createOrg: `, JSON.stringify(respCreate));
+    log(`Create @ CW done`);
```
Comment (thomasyopes) [inline packages/api/src/external/commonwell/command/cw-organization/create-or-update-cw-organization.ts:75]: If not leftover, add to CQ?

File: packages/api/src/external/carequality/command/cq-organization/__tests__/organization-template.test.ts [hunk 20]
```diff
@@ -0,0 +1,6 @@
+// TODO Add tests
```
Comment (RamilGaripov) [inline packages/api/src/external/carequality/command/cq-organization/__tests__/organization-template.test.ts:1]: nit: add ticket number

File: packages/api/src/external/carequality/command/cq-organization/create-or-update-cq-organization.ts [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    //eslint-disable-next-line @typescript-eslint/no-explicit-any
+  } catch (error: any) {
+    const extra = {
+      orgOid: orgDetails.oid,
+      context: `cq.org.update`,
+      carequalityOrg,
+      error,
+    };
+    if (error.response?.status === 404) {
+      capture.message("Got 404 while updating Org @ CQ, creating it", { extra });
```
Comment (RamilGaripov) [inline packages/api/src/external/carequality/command/cq-organization/create-or-update-cq-organization.ts:79]: nit: if not logging the error, maybe it's good to leave a comment it's intentional?

File: packages/api/src/external/carequality/command/create-or-update-facility.ts [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+  const parentOrgOid = isObo ? metriportIntermediaryOid : metriportOid;
+  return {
+    cxId: org.cxId,
+    name: cqOrgName,
+    oid: facility.oid,
+    address: facility.data.address,
+    contactName: metriportCompanyDetails.name,
+    phone: metriportCompanyDetails.phone,
+    email: metriportEmailForCq,
+    active: facility.cqActive, // Confirm, it was org.cqActive before
```
Comment (RamilGaripov) [inline packages/api/src/external/carequality/command/create-or-update-facility.ts:55]: TODO? Has this been resolved?

File: packages/api/src/external/commonwell/command/cw-organization/update-cw-organization-and-metriport-entity.ts [hunk 23]
```diff
@@ -0,0 +1,44 @@
+import { isOboFacility } from "../../../../domain/medical/facility";
+import { FacilityModel } from "../../../../models/medical/facility";
+import { OrganizationModel } from "../../../../models/medical/organization";
+import { createOrUpdateCwOrganization } from "./create-or-update-cw-organization";
+import { getOrgOrFail } from "./get-cw-organization";
+
+// TODO Apply the same pattern as Carequality
```
Comment (RamilGaripov) [inline packages/api/src/external/commonwell/command/cw-organization/update-cw-organization-and-metriport-entity.ts:7]: nit: ticket number

File: packages/carequality-sdk/src/client/carequality-fhir.ts [hunk 24]
```diff
@@ -154,30 +149,23 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
...
     query.append("apikey", this.apiKey);
-    query.append("_format", JSON_FORMAT);
     query.append("_count", count.toString());
-    query.append("_start", start.toString());
-    if (!active) query.append("_active", active.toString());
-    oid && query.append("_id", oid);
-    const queryString = query.toString();
+    start != undefined && query.append("_start", start.toString());
+    oid != undefined && query.append("_id", oid);
+    active != undefined && query.append("active", active.toString());
```
Comment (RamilGaripov) [inline packages/carequality-sdk/src/client/carequality-fhir.ts:163]: are we sure that the `active` query param doesnt need the underscore anymore?

File: packages/carequality-sdk/src/client/carequality-fhir.ts [hunk 25]
```diff
@@ -203,13 +190,12 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
    * @param oid string containing the organization OID
    * @returns an XML string containing an OperationOutcome resource - see Carequality documentation for details - https://carequality.org/healthcare-directory/OperationOutcome-create-success-example2.xml.html
    */
-  async updateOrganization(org: string, oid: string): Promise<string> {
+  async updateOrganization({ org, oid }: UpdateOrganizationParams): Promise<Organization> {
     const query = new URLSearchParams();
     query.append("apikey", this.apiKey);
-    query.append("_format", XML_FORMAT);
```
Comment (RamilGaripov) [inline packages/carequality-sdk/src/client/carequality-fhir.ts:209]: IIUC, we're removing this because default is json? 
Would it be better to set that explicitly? In the past, xml was the default, and then CQ changed it, causing a disruption on our end

File: packages/carequality-sdk/src/client/carequality-fhir.ts [hunk 26]
```diff
@@ -203,13 +190,12 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
...
 
-    const url = `${CarequalityManagementAPIImpl.ORG_ENDPOINT}/${oid}?${query.toString()}`;
-    const resp = await this.sendPutRequest(url, org, { "Content-Type": "application/xml" });
+    const url = `${CarequalityManagementAPIImplFhir.ORG_ENDPOINT}/${oid}?${query.toString()}`;
+    const resp = await this.sendPutRequest(url, org, { "Content-Type": "application/json" });
```
Comment (RamilGaripov) [inline packages/carequality-sdk/src/client/carequality-fhir.ts:198]: since all three functions are now using the same content type:
`"Content-Type": "application/json"`, 

let's just set in on the `buildHeaders` function

File: packages/carequality-sdk/src/client/carequality-api.ts
Comment (RamilGaripov) [inline packages/carequality-sdk/src/client/carequality-api.ts:1]: if we're removing the STU3 CQ API, maybe we should also remove all the schemas under `packages/carequality-sdk/src/models`?

File: packages/api/src/external/carequality/command/cq-directory/get-cq-directory-entry-as-organization.ts [hunk 27]
```diff
@@ -107,9 +111,34 @@ export async function getStandaloneOrganizations(): Promise<CQDirectoryEntryMode
...
+            },
+          ],
+        },
       },
-    },
+    })
+  ).map(org => org.dataValues);
+}
+
+function sortOrgsByNearbyOrder(orgs: CQDirectoryEntry[], orderMap: Map<string, number>) {
```
Comment (thomasyopes) [inline packages/api/src/external/carequality/command/cq-directory/get-cq-directory-entry-as-organization.ts:122]: Moved without edit from.... TODO

File: packages/api/src/external/carequality/command/cq-directory/get-cq-directory-entry-as-organization.ts [hunk 28]
```diff
@@ -107,9 +111,34 @@ export async function getStandaloneOrganizations(): Promise<CQDirectoryEntryMode
...
+  return orgs.sort((a, b) => {
+    const orderA = orderMap.get(a.id) ?? Number.MAX_SAFE_INTEGER;
+    const orderB = orderMap.get(b.id) ?? Number.MAX_SAFE_INTEGER;
+
+    return orderA - orderB;
   });
 }
+
+export async function getOrganizationsForXCPD(
```
Comment (thomasyopes) [inline packages/api/src/external/carequality/command/cq-directory/get-cq-directory-entry-as-organization.ts:131]: Moved without edit from.... TODO

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-facility.ts [hunk 29]
```diff
@@ -1,54 +0,0 @@
...
-import { metriportCompanyDetails } from "@metriport/shared";
-import { Facility, isOboFacility } from "../../../../domain/medical/facility";
-import { createOrUpdateCQOrganization } from "./create-or-update-cq-organization";
-import { buildCqOrgNameForFacility, getCqAddress } from "../../shared";
-import { Config } from "../../../../shared/config";
-
-export const metriportOid = Config.getSystemRootOID();
-export const metriportIntermediaryOid = `${metriportOid}.666`;
-
-export async function createOrUpdateFacilityInCq({
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-facility.ts:12]: Refactored into [createOrUpdateFacility](https://github.com/metriport/metriport/pull/2974/files#diff-ad4f53e95a1cf5a5613dea13d3894a7dc5b8ddd011d7adc0fbaf219fb56d87eeR21)

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts [hunk 30]
```diff
@@ -1,152 +0,0 @@
...
-import { makeCarequalityManagementAPI } from "../../api";
-import { CQOrganization } from "../../organization";
-import { CQOrgDetails, getCqAddress, getParsedCqOrgOrFail } from "../../shared";
-import { metriportIntermediaryOid, metriportOid } from "./create-or-update-cq-facility";
-import { metriportEmail as metriportEmailForCq } from "../../constants";
-import { metriportCompanyDetails } from "@metriport/shared";
-
-const cq = makeCarequalityManagementAPI();
-
-export async function createOrUpdateCQOrganization(
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts:18]: Refactored this into [createOrUpdateCQOrganization](https://github.com/metriport/metriport/pull/2974/files#diff-e9818b17b0e85fbe1d8c2d0c76bef56c6d392e024b71864e81e53c7d32eedd37R25)

File: packages/api/src/external/carequality/command/cq-organization/create-or-update-cq-organization.ts (hunk 21 above)
Comment (leite08) [reply packages/api/src/external/carequality/command/cq-organization/create-or-update-cq-organization.ts:79]: I think if it's important enough to send an event to Sentry, it's defs important enough to be logged.

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts (hunk 2 above)
Comment (leite08) [reply packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts:20]: This is now only being passed as part of the org's payload.

Gotta test if we can make an inactive CQ dir entry active by sending an update w/ this flag set to true. Otherwise, we'll prob need to adjust the logic or cancel the operation if the org exists but it's inactive.

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-organization.ts:1]: Refactored this into [createOrUpdateCQOrganization](https://github.com/metriport/metriport/pull/2974/files#diff-e9818b17b0e85fbe1d8c2d0c76bef56c6d392e024b71864e81e53c7d32eedd37R25) and [createOrUpdateCqOrganization](https://github.com/metriport/metriport/pull/2974/files#diff-e9818b17b0e85fbe1d8c2d0c76bef56c6d392e024b71864e81e53c7d32eedd37R25) (that's now reused across Org and Facility code).

File: packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-facility.ts
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/create-or-update-cq-facility.ts:1]: Refactored into [createOrUpdateFacility](https://github.com/metriport/metriport/pull/2974/files#diff-ad4f53e95a1cf5a5613dea13d3894a7dc5b8ddd011d7adc0fbaf219fb56d87eeR21) and [createOrUpdateCqOrganization](https://github.com/metriport/metriport/pull/2974/files#diff-e9818b17b0e85fbe1d8c2d0c76bef56c6d392e024b71864e81e53c7d32eedd37R25) (that's now reused across Org and Facility code).

File: packages/api/src/external/carequality/command/cq-directory/get-organizations-for-xcpd.ts [hunk 31]
```diff
@@ -1,30 +0,0 @@
...
-function sortOrgsByNearbyOrder(orgs: CQDirectoryEntryModel[], orderMap: Map<string, number>) {
-  return orgs.sort((a, b) => {
-    const orderA = orderMap.get(a.id) ?? Number.MAX_SAFE_INTEGER;
-    const orderB = orderMap.get(b.id) ?? Number.MAX_SAFE_INTEGER;
-
-    return orderA - orderB;
-  });
-}
-
-export async function getOrganizationsForXCPD(
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/get-organizations-for-xcpd.ts:17]: Moved to https://github.com/metriport/metriport/pull/2974/files#diff-4d528d0967a22a55b28f15970217a7615b95e02894fa6cd0302611827666a396R131

File: packages/api/src/external/carequality/command/cq-directory/get-organizations-for-xcpd.ts [hunk 32]
```diff
@@ -1,30 +0,0 @@
-import { CQDirectoryEntryModel } from "../../models/cq-directory";
-import {
-  getRecordLocatorServiceOrganizations,
-  getStandaloneOrganizations,
-  getSublinkOrganizations,
-} from "./cq-gateways";
-
-function sortOrgsByNearbyOrder(orgs: CQDirectoryEntryModel[], orderMap: Map<string, number>) {
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/get-organizations-for-xcpd.ts:8]: Moved to https://github.com/metriport/metriport/pull/2974/files#diff-4d528d0967a22a55b28f15970217a7615b95e02894fa6cd0302611827666a396R122

File: packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts [hunk 33]
```diff
@@ -1,224 +0,0 @@
...
-};
-
-const EARTH_RADIUS = 6378168;
-const XCPD_STRING = "ITI-55";
-const XCA_DQ_STRING = "ITI-38";
-const XCA_DR_STRING = "ITI-39";
-const XDR_STRING = "ITI-41";
-type ChannelUrl = typeof XCPD_STRING | typeof XCA_DQ_STRING | typeof XCA_DR_STRING;
-
-export function parseCQDirectoryEntries(orgsInput: Organization[]): CQDirectoryEntryData[] {
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts:28]: Refactored to https://github.com/metriport/metriport/pull/2974/files#diff-59870e46f472301c4f3bf527b5a922ed8f32fb3bd3eeff160e5e679be49c76ddR25

File: packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+import { transactionUrl } from "./organization-template";
+
+const EARTH_RADIUS = 6378168;
+const XCPD_STRING = "ITI-55";
+const XCA_DQ_STRING = "ITI-38";
+const XCA_DR_STRING = "ITI-39";
+const XDR_STRING = "ITI-41";
+type ChannelUrl = typeof XCPD_STRING | typeof XCA_DQ_STRING | typeof XCA_DR_STRING;
+
+export async function parseCQOrganization(org: Organization): Promise<CQDirectoryEntryData> {
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts:25]: Refactored from https://github.com/metriport/metriport/pull/2974/files#diff-f8c35511599e6dbd1bb86eb81dc9b2fee8e7a109456cc525e250b817795dee40L28

File: packages/api/src/external/carequality/organization-conversion.ts [hunk 35]
```diff
@@ -1,26 +0,0 @@
-import { uuidv7 } from "@metriport/core/util/uuid-v7";
-import { XCPDGateway } from "@metriport/ihe-gateway-sdk";
-import { CQOrgBasicDetails } from "./command/cq-directory/search-cq-directory";
-
-export async function cqOrgsToXCPDGateways(cqOrgs: CQOrgBasicDetails[]): Promise<XCPDGateway[]> {
```
Comment (leite08) [inline packages/api/src/external/carequality/organization-conversion.ts:5]: Moved w/o changes to https://github.com/metriport/metriport/pull/2974/files#diff-89a78675276350d64a8435988dc39d01941058d7d1d40f525938aae273593d42R79

File: packages/api/src/external/carequality/gateway/index.ts [hunk 36]
```diff
@@ -74,3 +75,26 @@ async function getE2eGateways(): Promise<XCPDGateway[]> {
   });
   return [e2eXcpdGateway];
 }
+
+async function cqOrgsToXCPDGateways(cqOrgs: CQOrgBasicDetails[]): Promise<XCPDGateway[]> {
```
Comment (leite08) [inline packages/api/src/external/carequality/gateway/index.ts:79]: Moved w/o changes from https://github.com/metriport/metriport/pull/2974/files#diff-06d0564d74cc17815c972ab145bb24a26758872b0850763259ce43a1eaab2c39L5

File: packages/api/src/external/carequality/gateway/index.ts [hunk 37]
```diff
@@ -74,3 +75,26 @@ async function getE2eGateways(): Promise<XCPDGateway[]> {
...
+        urlXCPD: org.urlXCPD,
+        id: org.id,
+      });
+      v2Gateways.push(gateway);
+    }
+  }
+  return v2Gateways;
+}
+
+function buildXcpdGateway(org: { id: string; urlXCPD: string }): XCPDGateway {
```
Comment (leite08) [inline packages/api/src/external/carequality/gateway/index.ts:94]: Moved w/o changes from https://github.com/metriport/metriport/pull/2974/files#diff-06d0564d74cc17815c972ab145bb24a26758872b0850763259ce43a1eaab2c39L20

File: packages/api/src/external/carequality/organization-template.ts [hunk 38]
```diff
@@ -1,217 +0,0 @@
...
-  XCA_DR_STRING,
-  XCPD_STRING,
-} from "@metriport/carequality-sdk/common/util";
-import { Config } from "../../shared/config";
-import { CQOrgDetailsWithUrls } from "./shared";
-
-const metriportOid = Config.getSystemRootOID();
-const metriportName = Config.getSystemRootOrgName();
-
-export function buildXmlStringFromTemplate(orgDetails: CQOrgDetailsWithUrls) {
```
Comment (leite08) [inline packages/api/src/external/carequality/organization-template.ts:13]: No more XML 🪦

File: packages/api/src/external/carequality/organization.ts
Comment (leite08) [inline packages/api/src/external/carequality/organization.ts:1]: Removed altogether

File: packages/api/src/external/carequality/command/create-or-update-facility.ts (hunk 22 above)
Comment (leite08) [reply packages/api/src/external/carequality/command/create-or-update-facility.ts:55]: Yes, this was a mistake from my end when I added that comment. The original code did use facility.cqActive ([reference](https://github.com/metriport/metriport/pull/2974/files#diff-5e25fdc76944273a04fbc066f5eaece27dc3d590722efaa6f315bdc42790c3b8L49)).

File: packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts (hunk 18 above)
Comment (leite08) [reply packages/api/src/external/carequality/command/cq-organization/parse-cq-organization.ts:37]: I think its good

File: packages/carequality-sdk/src/client/carequality-fhir.ts (hunk 25 above)
Comment (leite08) [reply packages/carequality-sdk/src/client/carequality-fhir.ts:209]: - apiKey: now its set as a header
- format: agreed
